File: lab-redes/python/multicast/cliente_multicast.py
import socket
import struct

# Parte C - Multicast (cliente/receptor) em Python. Entra no grupo e escuta.
# OFFSET = mesmo valor do servidor (ver README).
OFFSET = 39

GRUPO_MULTICAST = "230.0.0.1"
PORTA = 4446 + OFFSET  # 4485

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
# SO_REUSEADDR: deixa varios clientes na mesma maquina ouvirem a mesma porta
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(("", PORTA))

# entra no grupo multicast (IP_ADD_MEMBERSHIP) em todas as interfaces (INADDR_ANY)
grupo = socket.inet_aton(GRUPO_MULTICAST)
solicitacao_membro = struct.pack("4sL", grupo, socket.INADDR_ANY)
sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, solicitacao_membro)

print(f"[Multicast] Inscrito no grupo {GRUPO_MULTICAST}:{PORTA}. Aguardando avisos...")
while True:
    dados, endereco = sock.recvfrom(1024)
    print(f"[Multicast] Recebido: {dados.decode('utf-8')}")

File: lab-redes/java/multicast/ClienteMulticast.java
import java.net.*;
import java.io.IOException;

// Parte C - Multicast (cliente/receptor). Ele "entra" (joinGroup) no grupo e
// passa a receber tudo que for mandado pra aquele grupo:porta. Abra 2 ou 3
// clientes ao mesmo tempo pra ver todos recebendo o mesmo aviso.
public class ClienteMulticast {
    // OFFSET = mesmo valor do servidor (ver README)
    static final int OFFSET = 39;

    public static void main(String[] args) throws IOException {
        String grupoMulticast = "230.0.0.1";
        int porta = 4446 + OFFSET; // 4485

        try (MulticastSocket socket = new MulticastSocket(porta)) {
            InetAddress grupo = InetAddress.getByName(grupoMulticast);
            InetSocketAddress endpointGrupo = new InetSocketAddress(grupo, porta);

            // Em Wi-Fi corporativa/VPN o multicast costuma ser bloqueado. Pra testar
            // servidor e cliente na MESMA maquina, use a interface de loopback:
            // NetworkInterface interfaceRede = NetworkInterface.getByInetAddress(InetAddress.getLoopbackAddress());
            NetworkInterface interfaceRede = NetworkInterface.getByInetAddress(InetAddress.getLocalHost());

            socket.joinGroup(endpointGrupo, interfaceRede);
            System.out.println("[Multicast] Inscrito no grupo " + grupoMulticast + ":" + porta + ". Aguardando avisos...");

            byte[] buffer = new byte[1024];
            while (true) {
                DatagramPacket pacote = new DatagramPacket(buffer, buffer.length);
                socket.receive(pacote);
                String mensagem = new String(pacote.getData(), 0, pacote.getLength());
                System.out.println("[Multicast] Recebido: " + mensagem);
            }
        }
    }
}
